File: lint-arwaky/src-python/infrastructure/mcp_server_resources.py
"""MCP Server Resources — Resource handling for auto-linter."""

from __future__ import annotations

from pathlib import Path
from pydantic import AnyUrl
from mcp.types import Resource


def build_resources(project_root: Path) -> list[Resource]:
    """Build MCP Resource URIs for rule definitions and config."""
    resources: list[Resource] = []

    # Rule documentation resources
    docs_dir = project_root / "docs"
    if docs_dir.exists():
        for md_file in sorted(docs_dir.glob("*.md")):
            resources.append(
                Resource(
                    uri=AnyUrl(f"auto-linter://rules/{md_file.name}"),
                    name=f"Auto-Linter Rules: {md_file.stem}",
                    description=f"Rule definitions from {md_file.name}",
                    mimeType="text/markdown",
                )
            )

    # Config resource
    for config_name in [
        "auto_linter.config.python.yaml",
        "auto_linter.config.javascript.yaml",
        "auto_linter.config.rust.yaml",
        "auto_linter.config.json",
        "pyproject.toml",
    ]:
        config_path = project_root / config_name
        if config_path.exists():
            resources.append(
                Resource(
                    uri=AnyUrl(f"auto-linter://config/{config_name}"),
                    name=f"Auto-Linter Config: {config_name}",
                    description=f"Configuration from {config_name}",
                    mimeType="application/x-yaml"
                    if config_name.endswith(".yaml")
                    else "application/toml",
                )
            )

    return resources


async def read_resource(uri: str, project_root: Path) -> str:
    """Read a resource by URI."""
    if not uri.startswith("auto-linter://"):
        raise ValueError(f"Unknown resource URI scheme: {uri}")

    # Parse: auto-linter://rules/file.md or auto-linter://config/file.yaml
    parts = uri.removeprefix("auto-linter://").split("/", 1)
    if len(parts) != 2:
        raise ValueError(f"Invalid resource URI: {uri}")

    resource_type, filename = parts

    if resource_type == "rules":
        file_path = project_root / "docs" / filename
    elif resource_type == "config":
        file_path = project_root / filename
    else:
        raise ValueError(f"Unknown resource type: {resource_type}")

    # Security: ensure resolved path is under project root
    resolved = file_path.resolve()
    if not resolved.is_relative_to(project_root.resolve()):
        raise ValueError(f"Resource path escapes project root: {filename}")

    if not file_path.exists():
        raise FileNotFoundError(f"Resource not found: {filename}")

    return file_path.read_text(encoding="utf-8")

File: lint-arwaky/src-rust/infrastructure/mcp_server_resources.rs
/// mcp_server_resources — Resource handling for auto-linter.

#[derive(Debug, Clone)]
pub struct McpResource {
    pub uri: String,
    pub name: String,
    pub description: String,
    pub mime_type: String,
}

impl McpResource {
    pub fn new(uri: String, name: String, description: String, mime_type: String) -> Self {
        Self { uri, name, description, mime_type }
    }
}

pub fn build_resources(project_root: &str) -> Vec<McpResource> {
    let mut resources = Vec::new();
    let root = std::path::Path::new(project_root);

    let docs_dir = root.join("docs");
    if docs_dir.exists() && docs_dir.is_dir() {
        if let Ok(entries) = std::fs::read_dir(&docs_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().map(|e| e == "md").unwrap_or(false) {
                    let name = path.file_stem().unwrap_or_default().to_string_lossy().to_string();
                    resources.push(McpResource::new(
                        format!("auto-linter://rules/{}.md", name),
                        format!("Auto-Linter Rules: {}", name),
                        format!("Rule definitions from {}.md", name),
                        "text/markdown".to_string(),
                    ));
                }
            }
        }
    }

    for config_name in &["auto_linter.config.python.yaml", "auto_linter.config.json", "pyproject.toml"] {
        let config_path = root.join(config_name);
        if config_path.exists() {
            let mime = if config_name.ends_with(".yaml") { "application/x-yaml" } else if config_name.ends_with(".toml") { "application/toml" } else { "application/json" };
            resources.push(McpResource::new(
                format!("auto-linter://config/{}", config_name),
                format!("Auto-Linter Config: {}", config_name),
                format!("Configuration from {}", config_name),
                mime.to_string(),
            ));
        }
    }
    resources
}

pub fn read_resource(uri: &str, project_root: &str) -> Result<String, String> {
    if !uri.starts_with("auto-linter://") {
        return Err(format!("Unknown resource URI scheme: {}", uri));
    }
    let trimmed = uri.trim_start_matches("auto-linter://");
    let parts: Vec<&str> = trimmed.splitn(2, '/').collect();
    if parts.len() != 2 {
        return Err(format!("Invalid resource URI: {}", uri));
    }
    let resource_type = parts[0];
    let filename = parts[1];
    let root = std::path::Path::new(project_root);
    let file_path = if resource_type == "rules" {
        root.join("docs").join(filename)
    } else if resource_type == "config" {
        root.join(filename)
    } else {
        return Err(format!("Unknown resource type: {}", resource_type));
    };
    if !file_path.exists() {
        return Err(format!("Resource not found: {}", filename));
    }
    std::fs::read_to_string(&file_path).map_err(|e| format!("Failed to read resource: {}", e))
}
